src/osgPlugins/dds/ReaderWriterDDS.cpp

```diff
@@ -1010,7 +1027,28 @@ osg::Image* ReadDDSFile(std::istream& _istream, bool flipDDSRead)
         // this memory will not be used but it will not cause leak in worst meaning of this word.
     }
 
-    osgImage->setImage(s,t,r, internalFormat, pixelFormat, dataType, imageData, osg::Image::USE_NEW_DELETE, packing);
+    if (ddsd.ddpfPixelFormat.dwFlags & DDPF_PALETTEINDEXED8)
+    {
+        // Now we need to substitute the indexed image data with full RGBA image data.
+        unsigned char* convertedData = new unsigned char [sizeWithMipmaps * 4];
+        for (unsigned int i = 0; i < sizeWithMipmaps; i++)
+        {
+            convertedData[i * 4 + 0] = palette[imageData[i] * 4 + 0];
+            convertedData[i * 4 + 1] = palette[imageData[i] * 4 + 1];
+            convertedData[i * 4 + 2] = palette[imageData[i] * 4 + 2];
+            convertedData[i * 4 + 3] = palette[imageData[i] * 4 + 3];
+        }
```

Comment:
```
  unsigned char* convertedData = new unsigned char [sizeWithMipmaps * 4];
        unsigned char* pconvertedData = convertedData;
        unsigned char *pimageData = imageData;
        unsigned char *ppalet;
        for (unsigned int i = 0; i < sizeWithMipmaps; i++)
        {
ppalet=&palette[(*pimageData) * 4];
(*pconvertedData++) = *ppalet++;
(*pconvertedData++) = *ppalet++;
(*pconvertedData++) = *ppalet++;
++pimageData;
        }
```

Reply:
This seriously inhibits readability to questionable benefit. On my end, up to -O2 GCC either produces the exact same code or it's somehow even more complex with pointer arithmetics, while -O3 optimizes index-based loop even better than pointer-based loop. Not worth it.

Reply:
I'd never count on compiler to do what i want...most of the time it didn't...but it's your call

Reply:
Trying to outsmart the compiler isn't a brilliant idea.

Reply:
In most of cases compilers developers know C++ better than other developers, so usually you can prefer code readability over low-level optimizations.

Reply:
https://www.youtube.com/watch?v=x61H6qEtK08